python/src/iceberg/utils/schema_conversion.py

```diff
@@ -0,0 +1,451 @@
+
+    def _convert_array_type(self, array_type: Dict[str, Any]) -> ListType:
+        element_type, element_is_optional = self._resolve_inner_type(array_type, "items", "element-id")
+        return ListType(element_id=array_type["element-id"], element_type=element_type, element_is_optional=element_is_optional)
+
+    def _convert_map_type(self, map_type: Dict[str, Any]) -> MapType:
+        """
+        Args:
+            map_type: The dict that describes the Avro map type
+
+        Examples:
+            >>> from iceberg.utils.schema_conversion import AvroSchemaConversion
+            >>> avro_field = {
+            ...     "type": "map",
+            ...     "values": ["long", "null"],
+            ...     "key-id": 101,
+            ...     "value-id": 102,
+            ... }
+            >>> actual = AvroSchemaConversion()._convert_map_type(avro_field)
+            >>> expected = MapType(
+            ...     key_id=101,
+            ...     key_type=StringType(),
+            ...     value_id=102,
+            ...     value_type=LongType(),
+            ...     value_is_optional=True
+            ... )
+            >>> actual == expected
+            True
+
+        Returns: A MapType
+        """
+        value_type, value_is_optional = self._resolve_union(map_type["values"])
+        return MapType(
+            key_id=map_type["key-id"],
+            # Avro only supports string keys
+            key_type=StringType(),
+            value_id=map_type["value-id"],
+            value_type=self._convert_schema(value_type),
+            value_is_optional=value_is_optional,
+        )
+
+    def _convert_logical_type(self, avro_logical_type: Dict[str, Any]) -> IcebergType:
+        """
+        When a logical type is found, we'll resolve it here. For the decimal and map
+        we need to fetch more keys from the dict, and for the simple ones we can just
+        look it up in the mapping.
+
+        Examples:
+            >>> from iceberg.utils.schema_conversion import AvroSchemaConversion
+            >>> avro_logical_type = {
+            ...     "type": "int",
+            ...     "logicalType": "date"
+            ... }
+            >>> actual = AvroSchemaConversion()._convert_logical_type(avro_logical_type)
+            >>> actual == DateType()
+            True
+
+        Args:
+            avro_logical_type: The logical type
+
+        Returns:
+            The converted logical type
+
+        Raises:
+            ValueError: When the logical type is unknown
+        """
+        logical_type = avro_logical_type["logicalType"]
+        if logical_type == "decimal":
+            return self._convert_logical_decimal_type(avro_logical_type)
+        elif logical_type == "map":
+            return self._convert_logical_map_type(avro_logical_type)
+        elif logical_type in LOGICAL_FIELD_TYPE_MAPPING:
+            return LOGICAL_FIELD_TYPE_MAPPING[logical_type]
```

Comment:
Should we also validate the underlying type? The Java library does this automatically, but I don't think there's any validation in python that the actual type for a `date` is `int`.

Reply:
We can. I've added logic to lookup the logical/physical combination 👍🏻